### DISCUSSION TITLE
Adding in a Search Bar?

### DISCUSSION THREAD

**Discussion by SpencerPao** [alshedivat/al-folio#581] [Ideas]
Thoughts on adding a search bar at the top of the website? This can link to posts, announcements, etc.

**Comment by cmoralesmx** [comment]
Adding a search bar is dead simple. The challenging step is to actually integrate a search engine. 
Given that the project is a static-site-generator template, I see two options.
1. Implementing a user-side search engine such as [mini-search](https://github.com/lucaong/minisearch), [Fusejs](https://github.com/krisk/Fuse), or [Lunr.js](https://lunrjs.com/docs/index.html) which seems to be abandoned
2. Integrating a hosted solution such as [PSE](https://programmablesearchengine.google.com/about/), [meilisearch](https://www.meilisearch.com/), [search.io](https://www.search.io/), [algolia.com](https://www.algolia.com/)

Each option has tradeoffs that need to be discussed

**Reply by talker90** [reply]
I agree this would be a great addition. It looks like other popular themes use Algolia and Lunr.

e.g.
https://github.com/mmistakes/jekyll-theme-basically-basic

**Comment by xf-zhao** [comment]
Looking forward to any updates on this...

**Comment by george-gca** [comment]
Visually speaking I like solutions like [⌘K](https://cmdk.paco.me/) or [kbar](https://kbar.vercel.app/).

But about the search bar, is it necessary? I understand a search bar being useful in a documentation site, something with a lot more similar or related content that would be hard to find or filter in another way, but for a academic/portfolio site, at least for me, seems a bit unnecessary.

**Reply by xf-zhao** [reply]
I believe a search function helps if the author is active in writing blogs. Currently, we have the tags to help filter, but it only allows one tag to filter each time, it is hard to narrow down with tags. Then, either tags or a search function helps.

**Comment by george-gca** [comment]
Since Jekyll has no server side execution, we would have to rely on storing all the required content locally and search from there. If the blog posts increase a lot this would involve downloading the content of every post, so I don't think user-side search is the best option.

As for the hosted solution, [search.io](https://www.search.io/) was bought by [algolia.com](https://www.algolia.com/) and will be closed, [meilisearch](https://www.meilisearch.com/) free solution should be self hosted, which I believe doesn't work for our case. The remainder [PSE](https://programmablesearchengine.google.com/about/) and [algolia.com](https://www.algolia.com/) would need an account with a login, but these are the best options IMHO.

Edit: algolia have a [plugin for jekyll](https://github.com/algolia/jekyll-algolia), but it is not mantained anymore.

**Comment by george-gca** [comment]
Revisiting this discussion, I believe client side execution is the best one, at least for sites without too much content. Found a couple of templates ([just-the-docs](https://github.com/just-the-docs/just-the-docs) and [so-simple-theme](https://github.com/mmistakes/so-simple-theme)) that implemented this using [lunr.js](https://github.com/olivernn/lunr.js). But `lunr.js` haven't been updated in a while, so maybe a solution would be switch it for one of these:
- https://www.fusejs.io/
- https://github.com/nextapps-de/flexsearch
- https://github.com/bvaughn/js-search
- https://github.com/lucaong/minisearch

**Comment by george-gca** [comment]
Started to tackle this on #2415.
